File: Examples/Objective-C/BroadcastAlert/BroadcastAlert/SendNotificationViewController.m
//
//  SendNotificationViewController.m
//  BroadcastAlert
//
//  Created by Daniel Heredia on 7/27/16.
//  Copyright © 2017 Bridgefy Inc. All rights reserved.
//

#import "SendNotificationViewController.h"
#import "ReceivedNotificationsViewController.h"
#import <BFTransmitter/BFTransmitter.h>
#import <QuartzCore/QuartzCore.h>
#import "Constants.h"
#import <AudioToolbox/AudioServices.h>

#ifndef SEND_NOTIF
#define SEND_NOTIF
#define kSentKey @"sent_n"
#define kReceivedKey @"recev_n"
#endif


@interface SendNotificationViewController ()<BFTransmitterDelegate>
{
    NSInteger sentNumber;
    NSInteger receivedNumber;
    BOOL shouldVibrate;
}
@property (nonatomic, retain) BFTransmitter * transmitter;
@property (nonatomic, weak) ReceivedNotificationsViewController * receivedNotifsController;


@end

@implementation SendNotificationViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    //Transmitter initialization
    self.transmitter = [[BFTransmitter alloc] initWithApiKey:@"YOUR_API_KEY"];
    self.transmitter.delegate = self;
    [self.transmitter start];
    
    //UI controls load
    sentNumber = [self getSentNotificationsNumber];
    receivedNumber = [self getReceivedNotificationsNumber];
    self.nameLabel.text = [NSString stringWithFormat:@"%@", [[NSUserDefaults standardUserDefaults] stringForKey:USERNAME]];
    self.uuidLabel.text = [NSString stringWithFormat:@"%@", [self truncatedUUID]];
    [self refreshCounters];
    self.sentStatusLabel.text = @"";
    
    self.sendButton.layer.cornerRadius = 14.0;
    self.sendButton.layer.borderWidth = 2.0;
    self.sendButton.layer.borderColor = [UIColor redColor].CGColor;
    
    self.navigationItem.backBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@""
                                                                             style:UIBarButtonItemStylePlain
                                                                            target:nil
                                                                            action:nil];
    
    shouldVibrate = [[NSUserDefaults standardUserDefaults] boolForKey:VIBRATION_ENABLED];
    
    [self registerForNotifications];
}

-(void)refreshCounters
{
    self.sentNotificationsLabel.text =
    [NSString stringWithFormat:@"%ld", (long)sentNumber];
    self.receivedNotificationsLabel.text =
    [NSString stringWithFormat:@"%ld", (long)receivedNumber];
}

-(NSString *)truncatedUUID
{
    return [self.transmitter.currentUser substringToIndex:5];
}

- (void)registerForNotifications {
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(vibrationSettingChanged:)
                                                 name:VIBRATION_NOTIFICATION
                                               object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(usernameWasUpdated:)
                                                 name:USERNAME_NOTIFICATION
                                               object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(resetSentAlertsCounter:)
                                                 name:RESET_SENT_ALERTS_NOTIFICATION
                                               object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(deleteReceivedAlerts:)
                                                 name:DELETE_RECEIVED_ALERTS_NOTIFICATION
                                               object:nil];
}

#pragma mark - Navigation

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    
    if ([segue.identifier isEqualToString:@"received"])
    {
        self.receivedNotifsController = segue.destinationViewController;
    }
}


#pragma mark - Actions

-(IBAction)sendNotification:(id)sender
{
    //Sending the message.
    NSError * error;
    BFSendingOption options = (BFSendingOptionBroadcastReceiver |
                               BFSendingOptionNotEncrypted |
                               BFSendingOptionMeshTransmission);
    NSDictionary * dictionary = @{@"number": @(sentNumber + 1),
                                  @"date_sent": @(floor([[NSDate date] timeIntervalSince1970] * 1000)),
                                  @"device_name": [[NSUserDefaults standardUserDefaults] stringForKey:USERNAME]
                                  };
    [self.transmitter sendDictionary:dictionary
                              toUser:nil
                             options:options
                               error:&error];
    if (error)
    {
        NSLog(@"Error %@", error.localizedDescription);
    }
}


#pragma mark - BFTransmitterDelegate

- (void)transmitter:(BFTransmitter *)transmitter meshDidAddPacket:(NSString *)packetID
{
    // Packet added to mesh
    // Just called when the option BFSendingOptionMeshTransmission was used
    sentNumber++;
    self.sentStatusLabel.text = [NSString stringWithFormat:
                                 @"Broadcasting alert number %ld",
                                 (long)sentNumber];
    [self updateSentNotifications:sentNumber];
    [self refreshCounters];
}

- (void)transmitter:(BFTransmitter *)transmitter didReachDestinationForPacket:( NSString *)packetID
{
    //Mesh packet reached destiny (no always invoked)
}

- (void)transmitter:(BFTransmitter *)transmitter meshDidStartProcessForPacket:( NSString *)packetID;
{
    //A message entered in the mesh process (was added).
    // Just called when the option BFSendingOptionFullTransmission was used.
}

- (void)transmitter:(BFTransmitter *)transmitter didSendDirectPacket:(NSString *)packetID
{
    //A direct message was sent
}

- (void)transmitter:(BFTransmitter *)transmitter didFailForPacket:(NSString *)packetID error:(NSError * _Nullable)error
{
    //A direct message transmission failed.
}

- (void)transmitter:(BFTransmitter *)transmitter meshDidDiscardPackets:(NSArray<NSString *> *)packetIDs
{
    //A mesh message was discared and won't still be transmitted.
}

- (void)transmitter:(BFTransmitter *)transmitter meshDidRejectPacketBySize:(NSString *)packetID
{
    NSLog(@"The packet %@ was rejected from mesh because it exceeded the limit size.", packetID);
}

- (void)transmitter:(BFTransmitter *)transmitter
didReceiveDictionary:(NSDictionary<NSString *, id> * _Nullable) dictionary
           withData:(NSData * _Nullable)data
           fromUser:(NSString *)user
           packetID:(NSString *)packetID
          broadcast:(BOOL)broadcast
               mesh:(BOOL)mesh
{
    receivedNumber++;
    [self updateReceivedNotifications:receivedNumber];
    
    if (shouldVibrate) {
        AudioServicesPlaySystemSound(kSystemSoundID_Vibrate);
    }
    
    [self refreshCounters];
    // A dictionary was received by BFTransmitter.
    if (self.receivedNotifsController)
    {
        // If the the notifications screen is shown
        // update it.
        [self.receivedNotifsController addNotificationDictionary:dictionary
                                                        fromUUID:user];
    } else
    {
        //Otherwise, just update the data in file.
        [ReceivedNotificationsViewController addNotificationToFile:dictionary
                                                          fromUUID:user];
    }
    
}

- (void)transmitter:(BFTransmitter *)transmitter didDetectConnectionWithUser:(NSString *)user
{
    //A connection was detected (no necessarily secure)
}

- (void)transmitter:(BFTransmitter *)transmitter didDetectDisconnectionWithUser:(NSString *)user
{
    // A disconnection was detected.
}

- (void)transmitter:(BFTransmitter *)transmitter didFailAtStartWithError:(NSError *)error
{
    NSLog(@"An error occurred at start: %@", error.localizedDescription);
}

- (void)transmitter:(BFTransmitter *)transmitter didOccurEvent:(BFEvent)event description:(NSString *)description
{
    NSLog(@"Event reported: %@", description);
    
    if (event == BFEventStartFinished) {
        NSUserDefaults *usrDefaults = [NSUserDefaults standardUserDefaults];
        if (![usrDefaults boolForKey:INFO_SHOWED]) {
            [self infoPressed:nil];
            [usrDefaults setBool:YES
                          forKey:INFO_SHOWED];
            [usrDefaults synchronize];
        }
    }
}

- (BOOL)transmitter:(BFTransmitter *)transmitter shouldConnectSecurelyWithUser:(NSString *)user
{
    return NO; //if YES, it will establish connection with encryption capacities.
    // Not necessary for this case.
}

- (void)transmitter:(BFTransmitter *)transmitter didDetectSecureConnectionWithUser:(NSString *)user
{
    // A secure connection was detected
}

#pragma mark - Persistence of indicators
-(void)updateSentNotifications:(NSInteger)numNotifications
{
    [[NSUserDefaults standardUserDefaults] setObject:@(numNotifications) forKey:kSentKey];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

-(NSInteger)getSentNotificationsNumber
{
    NSNumber * value = [[NSUserDefaults standardUserDefaults] valueForKey:kSentKey];
    if (value == nil)
        return 0;
    return [value integerValue];
}
-(void)updateReceivedNotifications:(NSInteger)numNotifications
{
    [[NSUserDefaults standardUserDefaults] setObject:@(numNotifications) forKey:kReceivedKey];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

-(NSInteger)getReceivedNotificationsNumber
{
    NSNumber * value = [[NSUserDefaults standardUserDefaults] valueForKey:kReceivedKey];
    if (value == nil)
        return 0;
    return [value integerValue];
}

#pragma mark - Info method

- (IBAction)infoPressed:(id)sender {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Broadcast Alert"
                                                                             message:@"This app is designed to send and receive alerts without an internet connection."
                                                                      preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK"
                                                       style:UIAlertActionStyleDefault
                                                     handler:nil];
    
    [alertController addAction:okAction];
    
    [self presentViewController:alertController
                       animated:YES
                     completion:nil];
}

#pragma mark - Notification handlers

- (void)vibrationSettingChanged:(NSNotification *)notification {
    shouldVibrate = [[NSUserDefaults standardUserDefaults] boolForKey:VIBRATION_ENABLED];
}

- (void)usernameWasUpdated:(NSNotification *)notification {
    self.nameLabel.text = [NSString stringWithFormat:@"%@", [[NSUserDefaults standardUserDefaults] stringForKey:USERNAME]];
}

- (void)resetSentAlertsCounter:(NSNotification *)notification {
    sentNumber = 0;
    [self updateSentNotifications:sentNumber];
    [self refreshCounters];
    self.sentStatusLabel.text = @"";
}

- (void)deleteReceivedAlerts:(NSNotification *)notification {
    if ([ReceivedNotificationsViewController clearReceivedNotifications]) {
        receivedNumber = 0;
        [self updateReceivedNotifications:receivedNumber];
        [self refreshCounters];
    }
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}
@end

File: Examples/Swift/BroadcastAlert/BroadcastAlert/SendNotificationViewController.swift
//
//  SendNotificationViewController.swift
//  BroadcastAlert
//
//  Created by Daniel Heredia on 8/2/16.
//  Copyright © 2017 Bridgefy Inc. All rights reserved.
//

import UIKit
import BFTransmitter

let sent_key = "sent_n"
let received_key = "recev_n"


open class SendNotificationViewController: UIViewController, BFTransmitterDelegate {
    
    fileprivate var transmitter: BFTransmitter
    fileprivate weak var receivedNotifsController: ReceivedNotificationsViewController? = nil
    var sentNumber: Int
    var receivedNumber: Int
    
    //UI objects
    @IBOutlet weak var nameLabel: UILabel!
    @IBOutlet weak var uuidLabel: UILabel!
    @IBOutlet weak var sentNotificationsLabel: UILabel!
    @IBOutlet weak var sentStatusLabel: UILabel!
    @IBOutlet weak var receivedNotificationsLabel: UILabel!
    @IBOutlet weak var sendButton: UIButton!
    @IBOutlet weak var controlsContainer: UIView!
    
    public required init?(coder aDecoder: NSCoder) {
        
        //Transmitter initialization
        self.transmitter = BFTransmitter(apiKey: "YOUR API KEY")
        sentNumber = 0
        receivedNumber = 0
        super.init(coder: aDecoder)
        self.transmitter.delegate = self
        sentNumber = self.getSentNotificationsNumber()
        receivedNumber = self.getReceivedNotificationsNumber()
        
    }

    override open func viewDidLoad() {
        super.viewDidLoad()

        self.transmitter.start()
        //UI controls load
        self.nameLabel.text = "Device name: \(UIDevice.current.name)"
        self.uuidLabel.text = "User ID: \(self.truncatedUUID())"
        self.refreshCounters()
        self.sentStatusLabel.text = ""
        self.sendButton.layer.cornerRadius = 14.0
        self.sendButton.layer.borderWidth = 2.0
        self.sendButton.layer.borderColor = UIColor.red.cgColor
        self.controlsContainer.layer.cornerRadius = 14.0
    }
    
    func refreshCounters() {
        self.sentNotificationsLabel.text = "Sent alerts: \(sentNumber)"
        self.receivedNotificationsLabel.text = "Received alerts: \(receivedNumber)"
    }
    
    func truncatedUUID() -> String {
        let uuid: String = self.transmitter.currentUser!
        return uuid.substring(to: uuid.characters.index(uuid.startIndex, offsetBy: 5))
    }
    
    // MARK: Segue Method
    
    override open func prepare(for segue: UIStoryboardSegue, sender: Any!) {
        if segue.identifier == "received" {
            self.receivedNotifsController = segue.destination as? ReceivedNotificationsViewController
        }
    }
    
    // MARK: IB Actions
    
    @IBAction func sendNotification(_ sender: AnyObject) {
        //Sending the message.
        let dictionary: Dictionary<String, AnyObject> = ["number": (sentNumber + 1) as AnyObject,
                                                         "device_name": UIDevice.current.name as AnyObject,
                                                         "date_sent": floor(Date().timeIntervalSince1970 * 1000)  as AnyObject]
        let options: BFSendingOption = [.broadcastReceiver, .meshTransmission]
        
        do {
            try self.transmitter.send(dictionary, toUser: nil, options: options)
        }
        catch let err as NSError {
            print("Error: \(err)")
        }
    }
    
    // MARK: BFTransmitterDelegate
    
    public func transmitter(_ transmitter: BFTransmitter, meshDidAddPacket packetID: String) {
        // Packet added to mesh
        // Just called when the option BFSendingOptionMeshTransmission was used
        sentNumber += 1
        self.sentStatusLabel.text = "Alert number \(sentNumber) is being broadcasted!"
        self.updateSentNotifications(sentNumber)
        self.refreshCounters()
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didReachDestinationForPacket packetID: String) {
        //Mesh packet reached destiny (no always invoked)
    }
    
    public func transmitter(_ transmitter: BFTransmitter, meshDidStartProcessForPacket packetID: String) {
        //A message entered in the mesh process (was added).
        // Just called when the option BFSendingOptionFullTransmission was used.
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didSendDirectPacket packetID: String) {
        //A direct message was sent
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didFailForPacket packetID: String, error: Error?) {
        //A direct message transmission failed.
    }
    
    public func transmitter(_ transmitter: BFTransmitter, meshDidDiscardPackets packetIDs: [String]) {
        //A mesh message was discared and won't still be transmitted.

    }
    
    public func transmitter(_ transmitter: BFTransmitter, meshDidRejectPacketBySize packetID: String) {
        print("The packet \(packetID) was rejected from mesh because it exceeded the limit size.");
    }
    
    public func transmitter(_ transmitter: BFTransmitter,
        didReceive dictionary: [String : Any]?,
                    with data: Data?, fromUser
                         user: String,
                     packetID: String,
                    broadcast: Bool,
                         mesh: Bool) {
        receivedNumber += 1
        self.updateReceivedNotifications(receivedNumber)
        self.refreshCounters()
        // A dictionary was received by BFTransmitter.
        if self.receivedNotifsController != nil {
            // If the the notifications screen is shown
            // update it.
            self.receivedNotifsController!.addNotificationDictionary(dictionary!, fromUUID: user)
        } else {
            //Otherwise, just update the data in file.
            ReceivedNotificationsViewController.addNotificationToFile(dictionary!, fromUUID: user)
        }
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didDetectConnectionWithUser user: String) {
        //A connection was detected (no necessarily secure)
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didDetectDisconnectionWithUser user: String) {
        // A disconnection was detected.
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didFailAtStartWithError error: Error)
    {
        print("An error occurred at start: \(error.localizedDescription)");
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didOccur event: BFEvent, description: String)
    {
        print("Event reported: \(description)");
    }
    
    public func transmitter(_ transmitter: BFTransmitter, shouldConnectSecurelyWithUser user: String) -> Bool {
        return false//if True, it will establish connection with encryption capacities.
        // Not necessary for this case.
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didDetectSecureConnectionWithUser user: String) {
            // A secure connection was detected,
    }
    
    // MARK: Persistence of indicators
    
    func updateSentNotifications(_ numNotifications: Int) {
        UserDefaults.standard.set((numNotifications), forKey: sent_key)
        UserDefaults.standard.synchronize()
    }
    
    func getSentNotificationsNumber() -> Int {
        let value = UserDefaults.standard.value(forKey: sent_key)
        if value == nil {
            return 0
        }
        return (value! as AnyObject).intValue
    }
    
    func updateReceivedNotifications(_ numNotifications: Int) {
        UserDefaults.standard.set((numNotifications), forKey: received_key)
        UserDefaults.standard.synchronize()
    }
    
    func getReceivedNotificationsNumber() -> Int {
        let value = UserDefaults.standard.value(forKey: received_key)
        if value == nil {
            return 0
        }
        return (value! as AnyObject).intValue
    }

}
